book/scheduling-and-threading.md

```diff
@@ -312,256 +327,241 @@ situation---some devices have more [CPU cores][cores] than others, or are more
+}
 
-``` {.python}
-REFRESH_RATE_SEC = 0.016 # 16ms
+XMLHttpRequest.prototype.open = function(method, url, is_async) {
+    this.is_async = is_async
+    this.method = method;
+    this.url = url;
+}
+```
+
+* Add a new `__runXHROnload` method that will call the `onload` function
+specified on a `XMLHttpRequest`, if any.
+* Store the response on the `responseText` field, as required by the API.
+
+``` {.javascript}
+XMLHttpRequest.prototype.send = function(body) {
+    this.responseText = call_python("XMLHttpRequest_send",
+        this.method, this.url, this.body, this.is_async, this.handle);
+}
+
+function __runXHROnload(body, handle) {
+    var obj = XHR_REQUESTS[handle];
+    var evt = new Event('load');
+    obj.responseText = body;
+    if (obj.onload)
+        obj.onload(evt);
+}
 ```
 
-Asynchronous rendering
-======================
+And there you have it. With the task machinery and only a few more lines of
+non-plumbing code, we can support lots of new features, and `setTimeout` and
+async `XMLHttpRequest` are only the start.
 
-In order to separate rendering from other tasks and make it into a proper
-pipeline, let's make it asynchronous. Instead of updating rendering right away,
-the browser should set *dirty bits* (a fancy name for a boolean state variable)
-indicating that a particular part of the pipeline needs updating. Then when the
-pipeline is run, it will run the parts indicated by the dirty bits. We'll also
-need some way of
-*scheduling* the rendering pipeline to be updated at a given time in the
-future.
+Rendering pipeline tasks
+========================
 
-First, add one dirty bit to `Tab`:
+Everything in a browser can be considered a task, including rendering. In fact,
+the most important task in a browser is the rendering pipeline---but not just
+for the obvious reason that it's impossible to see web pages that aren't
+rendered. Most of the time spent doing work in a browser is in *rendering
+interactions* with the browser, such as loading, scrolling, clicking and
+typing. All of these interactions require rendering. If you want to make those
+interactions faster and smoother, the very first think you have to do is
+schedule the rendering pipeline, and to achieve that it'll need to be a
+schedulable task.
 
-* `needs_pipeline_update`, indicating that
-the pipeline needs to be re-run.
+On thing that is special about rendering is that it's a "singleton" task. There
+is only one rendering task, and either it's been scheduled or not, but it
+doesn't make sense to schedule it twice at the same time. On the other hand,
+it's totally fine and natural to have many `setTimeout` or `XMLHttpRequest`
+callbacks pending at the same time. The singleton nature of the rendering task
+has to do with there being only one DOM to render.
 
-And two to `TaskRunner`:
+Because it's a singleton, we'll need an additional boolean variable on a `Tab`
+to avoid having two rendering tasks, called `needs_animation_frame`. This
```

Comment:
This is a great explanation, but should we call it `has_rendering_task` or something like that instead of the current `needs_animation_frame`? Or we should drop in a footnote saying that the name will make a bit more sense later on.